### ISSUE TITLE
Simplify and streamline the Dockerfile

### ISSUE THREAD

**Issue by DeeDeeG** [RefugeRestrooms/refugerestrooms#509]
## Scope / difficulty

Just need to update the [`Dockerfile`](https://github.com/RefugeRestrooms/refugerestrooms/blob/develop/Dockerfile). Some discussion beforehand for consensus would be great.

## Impact

Would make `docker-compose build` run faster, and maybe make things easier to maintain; Should help the Docker config be more readable and easier to work with. Makes dependencies more obvious and more apparent what things are for what purpose.

## Rationale

I realized while working on a [branch](https://github.com/DeeDeeG/refugerestrooms/commits/vagrant-bionic-with-rvm-this-works-Oct-2018) that makes Vagrant usable (again, I'm not sure why I can't let go of Vagrant 🤷) that we can [simplify](https://github.com/DeeDeeG/refugerestrooms/commit/cb258ad3dbd9a71f206559340cf0938eec583acb) the Docker script. A lot of `apt-get update` and `apt-get install` in the script can be removed, now that we have the nodejs, phantomjs, and yarn install scripts that install most or all of those packages regardless.

I also noticed @btyy77c's branch over at https://github.com/btyy77c/refugerestrooms/tree/dockerAlpine, that shows how much neater the script can get. I think we might want to ask for some of those things as a Pull Request, or get the okay to re-implement some of those changes. (Would want to give credit in the commit messages.)

## Proposal

I think we should try some simplified `Dockerfile`s and discuss which is the best direction to go.

OR just make a pull request using some progress or improvements from the aforementioned branches:
- https://github.com/DeeDeeG/refugerestrooms/tree/cleanup-Dockerfile
- https://github.com/btyy77c/refugerestrooms/tree/dockerAlpine
And discuss there.

## How to actually do this:

- Consider improvements that can be made to the `Dockerfile`. See branches that are linked above for ideas.
- Probably merge #504.
  - It's a little late, but we could consider https://github.com/btyy77c/refugerestrooms/tree/dockerAlpine as an alternative to #504, but I think that would be unkind this late in review for that PR. Another downside is, we would be working with Alpine, which is a tad less familiar than Debian (which in turn is almost identical to working with Ubuntu). But we don't need to mess with the `Dockerfile` very often, so I don't think that's a big issue.
- Latstly, implement improvements by submitting the updated `Dockerfile` as a pull request.

OR

- Same as above, but start with the pull request and discuss and potentially make changes from there.

**Comment by DeeDeeG** [comment]
Examples of speed boosts: 

- After downloading the ruby `2.3.7`base image, our [`develop`](https://github.com/RefugeRestrooms/refugerestrooms/tree/develop) branch built in **_9 minutes_** and 1 second on my computer.

Whereas...

- After downloading the ruby `2.3.7-alpine-3.7` base image, this branch built in **_8 minutes_** and 4 seconds on my computer: https://github.com/btyy77c/refugerestrooms/tree/dockerAlpine
- After downloading the ruby `2.3.7-slim` base image, this branch built in **_13 minutes_** and 18 seconds on my computer: https://github.com/darpr/refugerestrooms/tree/slim-docker-image *
- After downloading the ruby `2.3.7` base image, this branch built in **_8 minutes_** and 16 seconds on my computer: https://github.com/DeeDeeG/refugerestrooms/tree/cleanup-Dockerfile
---
\* (The branch with the `2.3.7-slim` base image needs to download more packages than `2.3.7` non-slim, since it "slims down" by including less packages. Downloading and installing packages is slow. Still, this could be optimized by cleaning up the `Dockerfile`. This branch is equivalent to `develop` just with the slim ruby base image.)

**Comment by DeeDeeG** [comment]
Size of built images: 
```
refugerestrooms_web   latest      . . .      1.49GB   #  develop
<none>                <none>      . . .      1.48GB   #  cleanup-Dockerfile
<none>                <none>      . . .      1.16GB   #  slim-docker-image
<none>                <none>      . . .      789MB    #  dockerAlpine
```
Size of base images:
```
ruby                  2.3.7-alpine3.7     . . .     94.4MB   #  Alpine ruby base image
ruby                  2.3.7-slim          . . .     254MB    #  Slim (debian) ruby base image
ruby                  2.3.7               . . .     941MB    #  Regular (debian) ruby base image
postgres              latest              . . .     228MB    #  Postgres db base image
```

**Comment by DeeDeeG** [comment]
I have a branch where I updated the Dockerfile to be simpler, build a smaller image, and build faster. (This branch is still based on the [`ruby:2.5.3-slim` Docker image](https://hub.docker.com/_/ruby/)).

https://github.com/RefugeRestrooms/refugerestrooms/blob/update-Dockerfile/Dockerfile

The charts in the previous two comments are updated to show how it stacks up.

This is the second-fastest build time I've recorded out of the various Dockerfile configurations, at 7 minutes and 48 seconds. (Behind only the Alpine Linux-based image for Ruby 2.5.3.)

And it produces the third-smallest image in terms of disk space, at 827 MB. (Again, behind only the two Alpine Linux-based images.)

It still uses a larger base image than Alpine, though, at 179 MB for `ruby-2.5.3-slim`, vs 94.4 MB for `2.3.7-alpine3.7` or 45.5MB for `2.5.3-alpine3.7`. So it presumably takes a bit longer to download on slow or intermittent networks, or uses a bit more of a limited data plan.

I would say the slim (debian-based) image is fine, especially with the improved Dockerfile in [`update-Dockerfile`](https://github.com/RefugeRestrooms/refugerestrooms/blob/update-Dockerfile/Dockerfile), but the Alpine image would be somewhat better for faster build speed, and for lower network usage. As long as it doesn't cause issues of some kind, and so far I haven't seen anything come up.

That said, Debian is still more familiar than Alpine, and I would find it more reassuring to work with. So there is that trade-off.
